src/libasr/pass/intrinsic_function_registry.h

```diff
@@ -511,6 +812,207 @@ static inline ASR::asr_t* create_ListIndex(Allocator& al, const Location& loc,
+                logical_type, nullptr));
+            Vec<ASR::stmt_t *> if_body; if_body.reserve(al, 1);
+            Vec<ASR::expr_t *> tuple_ele; tuple_ele.reserve(al, 3);
+            {
+                ASR::ttype_t *char_type_0 = TYPE(ASR::make_Character_t(al, loc,
+                    1, 0, nullptr, nullptr, 0));
+                tuple_ele.push_back(al, args[0]);
+                tuple_ele.push_back(al, EXPR(ASR::make_StringConstant_t(al, loc,
+                    s2c(al, ""), char_type_0)));
+                tuple_ele.push_back(al, EXPR(ASR::make_StringConstant_t(al, loc,
+                    s2c(al, ""), char_type_0)));
+                Vec<ASR::ttype_t *> tuple_type_; tuple_type_.reserve(al, 1);
+                tuple_type_.push_back(al, char_type);
+                tuple_type_.push_back(al, char_type_0);
+                tuple_type_.push_back(al, char_type_0);
+                if_body.push_back(al, STMT(ASR::make_Assignment_t(al, loc,
+                    result, EXPR(ASR::make_TupleConstant_t(al, loc,
+                    tuple_ele.p, tuple_ele.n, TYPE(ASR::make_Tuple_t(al, loc,
+                    tuple_type_.p, tuple_type_.n)))), nullptr)));
+            }
+            Vec<ASR::stmt_t *> else_body; else_body.reserve(al, 1);
+            tuple_ele.p = nullptr; tuple_ele.n = 0;
+            tuple_ele.reserve(al, 3);
+            {
+                tuple_ele.push_back(al, EXPR(ASR::make_StringSection_t(al, loc,
+                    args[0], EXPR(ASR::make_IntegerConstant_t(al, loc, 0, int_type)),
+                    index, nullptr, char_type, nullptr)));
+                tuple_ele.push_back(al, args[1]);
+                tuple_ele.push_back(al, EXPR(ASR::make_StringSection_t(al, loc,
+                    args[0], EXPR(ASR::make_IntegerBinOp_t(al, loc, index,
+                        ASR::binopType::Add,
+                        EXPR(ASR::make_StringLen_t(al, loc, args[1], int_type, nullptr)),
+                        int_type, nullptr)),
+                    EXPR(ASR::make_StringLen_t(al, loc, args[0], int_type, nullptr)),
+                    nullptr, char_type, nullptr)));
+                else_body.push_back(al, STMT(ASR::make_Assignment_t(al, loc,
+                    result, EXPR(ASR::make_TupleConstant_t(al, loc,
+                    tuple_ele.p, tuple_ele.n, return_type)), nullptr)));
+            }
+            body.push_back(al, STMT(ASR::make_If_t(al, loc, a_test,
+                if_body.p, if_body.n, else_body.p, else_body.n)));
+            body.push_back(al, STMT(ASR::make_Return_t(al, loc)));
+        }
+        ASR::symbol_t *fn_sym = make_Function_t(fn_name, fn_symtab, dep, args,
+            body, return_var, Source, Implementation, nullptr);
+        scope->add_symbol(fn_name, fn_sym);
+        return EXPR(ASR::make_FunctionCall_t(al, loc, fn_sym, fn_sym,
+            new_args.p, new_args.n, return_type, compile_time_value, nullptr));
+    }
```

Comment:
This function is not "generic", correct? The signature is always the same.

It takes about 100 lines of ASR construction to do what can be done in about 11 lines of Python:

https://github.com/lcompilers/lpython/blob/aa1f1952676d66608b75b18c1c89c66978c87d88/src/runtime/lpython_builtin.py#L827-L838

I suggest we merge this, but why don't we try to figure out how to make these 100 lines become 11 lines?

Reply:
I have similar thoughts. The idea of moving functions from runtime library (`pure` surface language implementation) to the intrinsic function registry (C++ ASR implementation) is not very clear to me yet. What benefits does the transition bring us? (Or what is the motivation behind the transition?) Are we trying to maintain a unique copy of intrinsic functions for every frontend language? I would be appreciate it and be grateful if someone could clarify this.

Reply:
The main motivation for this is for generic functions like minloc() or any() that require very special loop structure based on the array rank and must work for any array type, so doing it here is less code than overloading 100+ functions in the front end language.

The secondary motivation is to have these implementations self-contained in libasr, and shared with lpython/lfortran (and all future frontends), given that they take less code than in the front end language.

The above case unfortunately takes more code than in the frontend language. So I see two ways forward:

* Either keep these in the frontend langauge (much easier to maintain, but will have to be duplicated for all frontends)
* Or figure out how to make the ASR as easy to write as the frontend language. I would say maximum 2x penalty in terms of lines of code, so 11 lines of Python could be maximum of 22 lines of C++, but ideally around 11 lines also. Either way 100 lines is too much.

I would like to see how to figure out the second option, as it would fix everything. To make progress on that, we need some kind of ASR Builder that has the most common defaults that effectively will allow to write code as if it was Python or Fortran.

CC @czgdp1807 and @rebcabin. This is a blocker. We don't need to figure it out this week, but we do need a solution say within a month or so.

Reply:
> * Either keep these in the frontend langauge (much easier to maintain, but will have to be duplicated for all frontends)

I think for common functions like `exp`, `sum`, `abs`, etc I won't write code for it in the frontend. It requires duplicating efforts in all the compilers and also increased code to be tested since each frontend will have code in its own syntax so higher chances of errors. By using intrinsic functions we implement and test only once so its only a one time pain which can be reduced with time. See my earlier comment https://github.com/lcompilers/lpython/pull/1673#issuecomment-1499936663 as well.

I can already see a lot of patterns here as pointed out in my latest review of this PR.

Reply:
I see no reason why we could not program in ASR as easily as in Python. Once we do, then the second approach is preferable, since it would be as easy to maintain, as in Python.

Here is the meat of the function:
```python
 res : tuple[str, str, str] 
 ind : i32 
 ind = _lpython_str_find(s, sep) 
 if ind == -1: 
     res = (s, "", "") 
 else:  
     res = (s[0:ind], sep, s[ind+len(sep): len(s)])     
 return res 
```
So how could this look like in C++ building ASR directly? Here is one idea:

```c++
auto res = Variable("res", Tuple({Str, Str, Str}))
auto ind = Variable("ind", i32)
body = {
    If(Eq(ind, i32(-1)), {
        Assignment(res, Tuple({s, "", ""});
    }, {
        Assignment(res, Tuple({StringSection(s, {0, ind}), sep, StringSection(s, {Add(ind, StringLen(sep)), StringLen(s))}));
    }),
    Assignment(result, res);
}
// Do something with "body"
```

The guiding principle: Let it look like the Clojure style S-expression of `--show-asr`, such as `(Assignment result res)`, but use the syntax of M-expressions: `Assignment(result, res)`. Let's use C++ variables like `res` and `ind` to store intermediate results. Let `Assignment` and other "heads" have the right defaults, so that one can use it like I did above. Under the hood, it probably needs to keep track of a few things (allocator, location info, the current active symbol table scope, etc.). Let's assume these things can be figured out. Wouldn't the above be very maintainable and readable?

Reply:
Yup, I had the same idea, I started implement one by one. 
`ASR::make_IntegerConstant` -> `i32(1)`
`ASR::make_Integer_t` -> i32 or int32

Reply:
Create a new namespace, like Builder, and just put all these things in.

Reply:
Yes. We just need to remove ASR's verbosity. I would like Builder to be a class because I don't want to keep passing `al, loc` to all the creator functions. Also, the ideas in [this comment](https://github.com/lcompilers/lpython/pull/1673#discussion_r1172122992) can be the base for builder API.

Reply:
Yes, let's make Builder a class and make the instance very simple, perhaps `b.Assignment(...)`, to make it very easy to type and visual. Yes, `al` should be hidden. The location should actually be the location in the C++ code I think. This would have to be done with a macro, to automatically infer the line number and filename. So maybe we need `Assignment` macro, not sure if it should be `ASSIGNMENT`, or `XAssignment`. Something short, but making sure it does not clash.

Reply:
Beautiful, it is now about 20 lines, but with different formatting it would be around 14 or so. This is very good, thanks!